server/src/main/java/org/candlepin/controller/OwnerProductShareManager.java

```diff
@@ -0,0 +1,247 @@
+     *                      upstream product
+     * @return recipient's ownerProduct if it exists, or resolve the most recently shared product from shared
+     * products if shares are requested to be included.
+     */
+    public List<Product> resolveProductsByIds(Owner recipientOwner, Collection<String> productIds,
+        boolean includeShares) {
+        Map<String, Product> result = new HashMap<String, Product>();
+        List<Product> existingProducts;
+        Set<String> missingProducts = new HashSet<String>();
+
+        if (productIds != null && !productIds.isEmpty()) {
+            // add all requested now so we can remove the found ones later.
+            missingProducts.addAll(productIds);
+            existingProducts = ownerProductCurator.getProductsByIds(recipientOwner, productIds).list();
+        }
+        else {
+            existingProducts = ownerProductCurator.getProductsByOwner(recipientOwner).list();
+        }
+
+        for (Product product : existingProducts) {
+            result.put(product.getId(), product);
+            missingProducts.remove(product.getId());
+        }
+
+        if (includeShares && CollectionUtils.isNotEmpty(missingProducts)) {
+            List<OwnerProductShare> shares = shareCurator.findProductSharesByRecipient(recipientOwner,
+                true, missingProducts);
+            for (OwnerProductShare share : shares) {
+                if (!result.containsKey(share.getId())) {
+                    result.put(share.getId(), share.getProduct());
+                }
+            }
+        }
+
+        return new ArrayList<Product>(result.values());
+    }
+
+    /**
+     * 1. fetch all existing ownerProducts for the recipient, and all existing share records for that
+     * recipient and product.
+     * 2. for all products that already exist, create a share record with active = false. The resolved
+     * product is recipient's pre-existing product
+     * 3. for all products that dont exist in recipient, the resolved product is the sharing org's product
+     *   3.1 if share from this owner already exists, update share_date and mark it active.
+     *   3.2 if share does not already exist from this owner, insert new share record.
+     *   3.3 if previously active product has a different uuid from the current product, mark the product
+     *   as 'refreshDue' so we can update all recipient owner's pools with this product and regenerate
+     *   respective ents. else we are done.
+     *
+     * @param sharingOwner the owner that is sharing the sharedProducts
+     * @param recipient the recipient owner of the share
+     * @param sharedProducts the products that are being shared from sharingOwner to recipient
+     * @return set of ResolvedProducts that also indicate if a refresh is required in the recipient for
+     * that owner
+     */
+    public Map<String, ResolvedProduct> resolveProductsAndUpdateProductShares(Owner sharingOwner,
```

Comment:
I would really like to see a more object-oriented approach here.  I have a general idea of what this code is meant to be doing and I'm having a hard time keeping this all in my head at once.

For example, resolving products and updating product shares seem, to me, to orthogonal to each other.  It is convenient to perform the actions at once, but logically speaking they're unrelated.  Could we break them apart?

Reply:
done